### ISSUE TITLE
bcast/allreduce refuse to process with rank-1 size-1 tensor with non one stride.

### ISSUE THREAD

**Issue by hzhangxyz** [mpi4py/mpi4py#238]
In pytorch, `torch.tensor([1.+2.j]).real.contiguous().stride()` returns `(2,)` instead of `(1,)`, but the data in memory is contiguous indeed because of its size is only one. while bcast/allreduce of mpi4py raise an error saying it is not contiguous because of stride mismatch.

also see: https://github.com/pytorch/pytorch/issues/82610, they say it is contiguous already so nothing more should be done.

**Comment by leofang** [comment]
We had a discussion just last week on cuDF/CuPy interop. For example,
```python
>>> cp.empty((0,), dtype=cp.complex128).real.strides  # cupy v10, which cuDF expects (in bytes)
(16,)
>>> cp.empty((0,), dtype=cp.complex64).real.strides  # cupy v10, which cuDF expects (in bytes)
(8,)
>>> cp.empty((0,), dtype=cp.complex128).real.strides  # numpy & cupy v11 (in bytes)
(0,)
>>> cp.empty((0,), dtype=cp.complex64).real.strides  # numpy & cupy v11 (in bytes)
(0,)
>>> cp.empty((), dtype=cp.complex64).real.strides  # strides = null
()
```
and now you mentioned PyTorch returns strides = 2 (in counts) = 16 or 32 (in bytes, depending on the precision). This is endless.

On the cuDF side, it's likely they'll relax the constraint. The way I see it, as long as there's no interop standard regulating how to compute strides for corner cases, each library (including mpi4py) can do whatever they want. 

I do plan to raise this discussion to DLPack (for `__dlpack__` that the Python Array API Standard uses) / Numba (for `__cuda_array_interface__`) but probably not today. Feel free to open an issue in either repo.

**Comment by dalcinl** [comment]
> ```python
> >>> # so far so good... how about these?
> >>> cp.empty((0,), dtype=cp.complex128).real.strides  # cupy v10, which cuDF expects (in bytes)
> (16,)
> >>> cp.empty((0,), dtype=cp.complex64).real.strides  # cupy v10, which cuDF expects (in bytes)
> (8,)
> ```

IMHO, this is clearly wrong. You want strides zero, that way any pointer arithmetic on the base pointer produces the base pointer again (particularly important if the base pointer is NULL and you want to preserve this).
BTW, does `np.empty((0,1,2), dtype=np.complex64).real.strides` return `(0,0,0)` ? IOW, if the size is zero (but some items in shape may be not), strides should all be zero.

> ```python
> >>> cp.empty((0,), dtype=cp.complex128).real.strides  # numpy & cupy v11 (in bytes)
> (0,)
> >>> cp.empty((0,), dtype=cp.complex64).real.strides  # numpy & cupy v11 (in bytes)
> (0,)
> >>> cp.empty((), dtype=cp.complex64).real.strides  # strides = null, numpy also does this
> ()
> ```

Oh, now I see. So things got fixed in cupy v11.

**Comment by dalcinl** [comment]
> each library (including mpi4py) can do whatever they want.

I'm OK with adding a fix for special casing `ndim==0` when  checking for C/F contiguous. But this fix would go to master, and it would be available in the next major mpi4py v4.0 release.

**Comment by dalcinl** [comment]
> I'm OK with adding a fix for special casing `ndim==0`

Oh, sorry, `ndim==0` is already handled properly in mpi4py. 

The real problem here is, in general, `product(shape) = size = 1` arrays. mpi4py should accept all of these as contiguous. Am I missing something?

**Comment by hzhangxyz** [comment]
> The real problem here is, in general, `product(shape) = size = 1` arrays. mpi4py should accept all of these as contiguous. Am I missing something?

It seems the stride value of the same index should be ignored if `any of element in shape = 1`, not only `product(shape) = 1`

**Comment by dalcinl** [comment]
> It seems the stride value of the same index should be ignored if `any of element in shape = 1`, not only `product(shape) = 1`

BTW, this exactly the way core CPython does it [[link]](https://github.com/python/cpython/blob/v3.11.0b5/Objects/abstract.c#L453-L456).

**Comment by dalcinl** [comment]
@hzhangxyz Any chance you can give a try to mpi4py branch `bugfix/buf-contig-check/maint`?

**Comment by wence-** [comment]
> BTW, does `np.empty((0,1,2), dtype=np.complex64).real.strides` return `(0,0,0)`

Lol no :P
```
In [6]: np.empty((1, 0, 2), dtype=float).strides
Out[6]: (16, 16, 8)

In [7]: np.empty((1, 0, 0), dtype=float).strides
Out[7]: (8, 8, 8)

In [8]: np.empty((0, 0, 0), dtype=float).strides
Out[8]: (8, 8, 8)

In [9]: np.empty((0, 0, 1), dtype=float).strides
Out[9]: (8, 8, 8)

In [10]: np.empty((0, 2, 1), dtype=float).strides
Out[10]: (16, 8, 8)

In [11]: np.empty((0, 2, 0), dtype=float).strides
Out[11]: (16, 8, 8)
```

**Comment by hzhangxyz** [comment]
> @hzhangxyz Any chance you can give a try to mpi4py branch `bugfix/buf-contig-check/maint`?

It works well in my situation. :)

**Comment by dalcinl** [comment]
Fixed via 9ccf55fb0a0218d3fcc5eb7292ed14117d781852.
